### ISSUE TITLE
Add option "required" to the config.from_* methods

### ISSUE THREAD

**Issue by rmk135** [ets-labs/python-dependency-injector#369]
> Re: ``config.from_yaml('not_existent.yaml')``
> I would usually expect to have an exception raised when trying to do something with non-existent resources.

I agree. This is a design error. It didn't look like that when I designed it initially. Sort of hindsight bias :)

I can not change ``config.from_yaml('not_existent.yaml')`` to raise an error cause it will change current behaviour. There might be people who rely on current behaviour and their code will crash.

To fix the issue I could introduce a new API. Something like:

- ``config.from_yaml('not_existent.yaml', required=True)``
- ``config.from_ini('not_existent.yaml', required=True)``
- ``config.from_env('UNDEFINED', required=True)``
- ``config.from_dict({}, required=True)``

What do you think?

_Originally posted by @rmk135 in https://github.com/ets-labs/python-dependency-injector/issues/358#issuecomment-763638517_

--------

> Using a _required_ flag seems a great option, though I think this should be passed with kwargs, as this would allow passing things like a custom yaml loader (I will open a new issue for a potential problem I spotted).

_Originally posted by @StefanoFrazzetto in https://github.com/ets-labs/python-dependency-injector/issues/358#issuecomment-764520364_

**Comment by StefanoFrazzetto** [comment]
In #358 I suggested adding kwargs to the *from_\** methods. This is because of *Configuration.from_yaml* using an [unsafe loader](https://github.com/yaml/pyyaml/wiki/PyYAML-yaml.load(input)-Deprecation)* on [this line](malicious.yml). This would allow to have a bit more flexibility, keeping the interface consistent for all the methods while also allowing users to make their own choice.

See the following example using dependency-injector 4.10.3
```python
# issue.py
from dependency_injector import providers

config = providers.Configuration()
config.from_yaml('malicious.yml')
```

```yaml
# malicious.yml
!!python/object/new:os.system [echo EXPLOIT!]
```

This should be fixed by changing the method as
```python
def from_yaml(self, filepath, **kwargs):
  ...
  try:
    with open(filepath) as opened_file:
      loader = kwargs.get('loader', yaml.SafeLoader)  # <--- fall back to the safer option
      config = yaml.load(opened_file, loader)
  ...
```

If you think SafeLoader might break existing code, then it might be worth considering `FullLoader` instead; this will at least print a warning, making users aware of this issue.

*I had to dig a bit to find this: [*Loader*](https://github.com/yaml/pyyaml/blob/master/lib3/yaml/loader.py#L41) and *UnsafeLoader* are basically the same, but the former was left there (for backward compatibility) with no added warnings 🤷‍♂️

**Comment by StefanoFrazzetto** [comment]
PS: Opened a pull request as well, so I won't just complain about issues 😆

**Comment by rmk135** [comment]
Created #372 from your comment.

**Comment by rmk135** [comment]
> PS: Opened a pull request as well, so I won't just complain about issues 😆

No worries about complaining: it's not a complaining, it's a feedback. I'm positive and thankful you provide it. It helps to improve the framework.

Thanks for the PR! I will decline it cause some more work should be done before merging. I'll add you as a watcher to the new one so you could take a look on the change and maybe come up with the other PRs later.

**Comment by rmk135** [comment]
@StefanoFrazzetto 

Done at ``4.11.0``. It's already on PyPI. See: https://python-dependency-injector.ets-labs.org/providers/configuration.html#mandatory-and-optional-sources

Thanks for the improvement suggestion.
